stratosphere/loader/source/ldr_capabilities.cpp

```diff
@@ -0,0 +1,428 @@
+
+                        if (restrict.GetInterruptId0() == EmptyInterruptId && restrict.GetInterruptId1() == EmptyInterruptId) {
+                            return true;
+                        }
+
+                        if (restrict.GetInterruptId0() == id || restrict.GetInterruptId1() == id) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                return IsSingleIdValid(this->GetInterruptId0(), kac, kac_count) && IsSingleIdValid(this->GetInterruptId1(), kac, kac_count);
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(ApplicationType,
+            u32 GetApplicationType() const {
+                return (this->GetValue() >> 0) & 0x3;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return restrict.GetValue() == this->GetValue();
+                    }
+                }
+                return false;
+            }
+
+            static constexpr u32 Encode(u32 app_type) {
+                return ((app_type & 3) << ValueShift) | IdMask;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(KernelVersion,
+            u32 GetMinorVersion() const {
+                return (this->GetValue() >> 0) & 0xF;
+            }
+
+            u32 GetMajorVersion() const {
+                /* TODO: Are upper bits unused? */
+                return (this->GetValue() >> 4) & 0x1FFF;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return restrict.GetValue() == this->GetValue();
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(HandleTable,
+            u32 GetSize() const {
+                return (this->GetValue() >> 0) & 0x3FF;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return this->GetSize() <= restrict.GetSize();
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(DebugFlags,
+            bool GetAllowDebug() const {
+                return ((this->GetValue() >> 0) & 1) != 0;
```

Comment:
I'm pretty sure `!= 0` isn't required in this context, especially when the result from the bitmasking expression is already 1 or 0.

Reply:
Yeah, changed.